File: binding/src/multimedia/libqmediatimerange.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qmediatimerange.html
pub const qmediatimerange = struct {

    /// New constructs a new QMediaTimeRange object.
    pub fn New() QtC.QMediaTimeRange {
        return qtc.QMediaTimeRange_new();
    }


    /// New2 constructs a new QMediaTimeRange object.
    pub fn New2(start: i64, end: i64) QtC.QMediaTimeRange {
        return qtc.QMediaTimeRange_new2(start, end);
    }


    /// New3 constructs a new QMediaTimeRange object.
    pub fn New3(param1: ?*anyopaque) QtC.QMediaTimeRange {
        return qtc.QMediaTimeRange_new3(@ptrCast(param1));
    }


    /// New4 constructs a new QMediaTimeRange object.
    pub fn New4(range: ?*anyopaque) QtC.QMediaTimeRange {
        return qtc.QMediaTimeRange_new4(@ptrCast(range));
    }


    pub fn OperatorAssign(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QMediaTimeRange_OperatorAssign(@ptrCast(self), @ptrCast(param1));
    }

    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QMediaTimeRange_Swap(@ptrCast(self), @ptrCast(other));
    }

    pub fn Detach(self: ?*anyopaque, ) void {
        qtc.QMediaTimeRange_Detach(@ptrCast(self));
    }

    pub fn OperatorAssign2(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QMediaTimeRange_OperatorAssign2(@ptrCast(self), @ptrCast(param1));
    }

    pub fn EarliestTime(self: ?*anyopaque, ) i64 {
        return qtc.QMediaTimeRange_EarliestTime(@ptrCast(self));
    }

    pub fn LatestTime(self: ?*anyopaque, ) i64 {
        return qtc.QMediaTimeRange_LatestTime(@ptrCast(self));
    }

    pub fn Intervals(self: ?*anyopaque, ) []const u8 {
        return qtc.QMediaTimeRange_Intervals(@ptrCast(self));
    }

    pub fn IsEmpty(self: ?*anyopaque, ) bool {
        return qtc.QMediaTimeRange_IsEmpty(@ptrCast(self));
    }

    pub fn IsContinuous(self: ?*anyopaque, ) bool {
        return qtc.QMediaTimeRange_IsContinuous(@ptrCast(self));
    }

    pub fn Contains(self: ?*anyopaque, time: i64) bool {
        return qtc.QMediaTimeRange_Contains(@ptrCast(self), time);
    }

    pub fn AddInterval(self: ?*anyopaque, start: i64, end: i64) void {
        qtc.QMediaTimeRange_AddInterval(@ptrCast(self), start, end);
    }

    pub fn AddInterval2(self: ?*anyopaque, interval: ?*anyopaque) void {
        qtc.QMediaTimeRange_AddInterval2(@ptrCast(self), @ptrCast(interval));
    }

    pub fn AddTimeRange(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QMediaTimeRange_AddTimeRange(@ptrCast(self), @ptrCast(param1));
    }

    pub fn RemoveInterval(self: ?*anyopaque, start: i64, end: i64) void {
        qtc.QMediaTimeRange_RemoveInterval(@ptrCast(self), start, end);
    }

    pub fn RemoveInterval2(self: ?*anyopaque, interval: ?*anyopaque) void {
        qtc.QMediaTimeRange_RemoveInterval2(@ptrCast(self), @ptrCast(interval));
    }

    pub fn RemoveTimeRange(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QMediaTimeRange_RemoveTimeRange(@ptrCast(self), @ptrCast(param1));
    }

    pub fn OperatorPlusAssign(self: ?*anyopaque, param1: ?*anyopaque) QtC.QMediaTimeRange {
        return qtc.QMediaTimeRange_OperatorPlusAssign(@ptrCast(self), @ptrCast(param1));
    }

    pub fn OperatorPlusAssign2(self: ?*anyopaque, param1: ?*anyopaque) QtC.QMediaTimeRange {
        return qtc.QMediaTimeRange_OperatorPlusAssign2(@ptrCast(self), @ptrCast(param1));
    }

    pub fn OperatorMinusAssign(self: ?*anyopaque, param1: ?*anyopaque) QtC.QMediaTimeRange {
        return qtc.QMediaTimeRange_OperatorMinusAssign(@ptrCast(self), @ptrCast(param1));
    }

    pub fn OperatorMinusAssign2(self: ?*anyopaque, param1: ?*anyopaque) QtC.QMediaTimeRange {
        return qtc.QMediaTimeRange_OperatorMinusAssign2(@ptrCast(self), @ptrCast(param1));
    }

    pub fn Clear(self: ?*anyopaque, ) void {
        qtc.QMediaTimeRange_Clear(@ptrCast(self));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QMediaTimeRange_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qmediatimerange-interval.html
pub const qmediatimerange__interval = struct {

    /// New constructs a new QMediaTimeRange::Interval object.
    pub fn New(other: ?*anyopaque) QtC.QMediaTimeRange__Interval {
        return qtc.QMediaTimeRange__Interval_new(@ptrCast(other));
    }


    /// New2 constructs a new QMediaTimeRange::Interval object.
    pub fn New2(other: ?*anyopaque) QtC.QMediaTimeRange__Interval {
        return qtc.QMediaTimeRange__Interval_new2(@ptrCast(other));
    }


    /// New3 constructs a new QMediaTimeRange::Interval object.
    pub fn New3() QtC.QMediaTimeRange__Interval {
        return qtc.QMediaTimeRange__Interval_new3();
    }


    /// New4 constructs a new QMediaTimeRange::Interval object.
    pub fn New4(start: i64, end: i64) QtC.QMediaTimeRange__Interval {
        return qtc.QMediaTimeRange__Interval_new4(start, end);
    }


    /// New5 constructs a new QMediaTimeRange::Interval object.
    pub fn New5(param1: ?*anyopaque) QtC.QMediaTimeRange__Interval {
        return qtc.QMediaTimeRange__Interval_new5(@ptrCast(param1));
    }

    /// CopyAssign shallow copies `other` into `self`.
    pub fn CopyAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QMediaTimeRange__Interval_CopyAssign(@ptrCast(self), @ptrCast(other));
    }

    /// MoveAssign moves `other` into `self` and invalidates `other`.
    pub fn MoveAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QMediaTimeRange__Interval_MoveAssign(@ptrCast(self), @ptrCast(other));
    }


    pub fn Start(self: ?*anyopaque, ) i64 {
        return qtc.QMediaTimeRange__Interval_Start(@ptrCast(self));
    }

    pub fn End(self: ?*anyopaque, ) i64 {
        return qtc.QMediaTimeRange__Interval_End(@ptrCast(self));
    }

    pub fn Contains(self: ?*anyopaque, time: i64) bool {
        return qtc.QMediaTimeRange__Interval_Contains(@ptrCast(self), time);
    }

    pub fn IsNormal(self: ?*anyopaque, ) bool {
        return qtc.QMediaTimeRange__Interval_IsNormal(@ptrCast(self));
    }

    pub fn Normalized(self: ?*anyopaque, ) QtC.QMediaTimeRange__Interval {
        return qtc.QMediaTimeRange__Interval_Normalized(@ptrCast(self));
    }

    pub fn Translated(self: ?*anyopaque, offset: i64) QtC.QMediaTimeRange__Interval {
        return qtc.QMediaTimeRange__Interval_Translated(@ptrCast(self), offset);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QMediaTimeRange__Interval_Delete(@ptrCast(self));
    }
};

File: binding/src/multimedia/libqmediatimerange.cpp
#include <QMediaTimeRange>
#define WORKAROUND_INNER_CLASS_DEFINITION_QMediaTimeRange__Interval
#include <qmediatimerange.h>
#include "libqmediatimerange.h"
#include "libqmediatimerange.hxx"

QMediaTimeRange* QMediaTimeRange_new() {
	 return new QMediaTimeRange();
}

QMediaTimeRange* QMediaTimeRange_new2(long long start, long long end) {
	 return new QMediaTimeRange(start, end);
}

QMediaTimeRange* QMediaTimeRange_new3(const QMediaTimeRange__Interval* param1) {
	 return new QMediaTimeRange(*param1);
}

QMediaTimeRange* QMediaTimeRange_new4(const QMediaTimeRange* range) {
	 return new QMediaTimeRange(*range);
}

void QMediaTimeRange_OperatorAssign(QMediaTimeRange* self, const QMediaTimeRange* param1) {
	self->operator=(*param1);
}

void QMediaTimeRange_Swap(QMediaTimeRange* self, QMediaTimeRange* other) {
	self->swap(*other);
}

void QMediaTimeRange_Detach(QMediaTimeRange* self) {
	self->detach();
}

void QMediaTimeRange_OperatorAssign2(QMediaTimeRange* self, const QMediaTimeRange__Interval* param1) {
	self->operator=(*param1);
}

long long QMediaTimeRange_EarliestTime(const QMediaTimeRange* self) {
	return self->earliestTime();
}

long long QMediaTimeRange_LatestTime(const QMediaTimeRange* self) {
	return self->latestTime();
}

libqt_list QMediaTimeRange_Intervals(const QMediaTimeRange* self) {
	auto _ret = self->intervals();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

bool QMediaTimeRange_IsEmpty(const QMediaTimeRange* self) {
	return self->isEmpty();
}

bool QMediaTimeRange_IsContinuous(const QMediaTimeRange* self) {
	return self->isContinuous();
}

bool QMediaTimeRange_Contains(const QMediaTimeRange* self, long long time) {
	return self->contains(time);
}

void QMediaTimeRange_AddInterval(QMediaTimeRange* self, long long start, long long end) {
	self->addInterval(start, end);
}

void QMediaTimeRange_AddInterval2(QMediaTimeRange* self, const QMediaTimeRange__Interval* interval) {
	self->addInterval(*interval);
}

void QMediaTimeRange_AddTimeRange(QMediaTimeRange* self, const QMediaTimeRange* param1) {
	self->addTimeRange(*param1);
}

void QMediaTimeRange_RemoveInterval(QMediaTimeRange* self, long long start, long long end) {
	self->removeInterval(start, end);
}

void QMediaTimeRange_RemoveInterval2(QMediaTimeRange* self, const QMediaTimeRange__Interval* interval) {
	self->removeInterval(*interval);
}

void QMediaTimeRange_RemoveTimeRange(QMediaTimeRange* self, const QMediaTimeRange* param1) {
	self->removeTimeRange(*param1);
}

QMediaTimeRange* QMediaTimeRange_OperatorPlusAssign(QMediaTimeRange* self, const QMediaTimeRange* param1) {
	return new QMediaTimeRange(self->operator+=(*param1));
}

QMediaTimeRange* QMediaTimeRange_OperatorPlusAssign2(QMediaTimeRange* self, const QMediaTimeRange__Interval* param1) {
	return new QMediaTimeRange(self->operator+=(*param1));
}

QMediaTimeRange* QMediaTimeRange_OperatorMinusAssign(QMediaTimeRange* self, const QMediaTimeRange* param1) {
	return new QMediaTimeRange(self->operator-=(*param1));
}

QMediaTimeRange* QMediaTimeRange_OperatorMinusAssign2(QMediaTimeRange* self, const QMediaTimeRange__Interval* param1) {
	return new QMediaTimeRange(self->operator-=(*param1));
}

void QMediaTimeRange_Clear(QMediaTimeRange* self) {
	self->clear();
}

void QMediaTimeRange_Delete(QMediaTimeRange* self) {
    delete self;
}

QMediaTimeRange__Interval* QMediaTimeRange__Interval_new(const QMediaTimeRange__Interval* other) {
	 return new QMediaTimeRange::Interval(*other);
}

QMediaTimeRange__Interval* QMediaTimeRange__Interval_new2(QMediaTimeRange__Interval* other) {
	 return new QMediaTimeRange::Interval(*other);
}

QMediaTimeRange__Interval* QMediaTimeRange__Interval_new3() {
	 return new QMediaTimeRange::Interval();
}

QMediaTimeRange__Interval* QMediaTimeRange__Interval_new4(long long start, long long end) {
	 return new QMediaTimeRange::Interval(start, end);
}

QMediaTimeRange__Interval* QMediaTimeRange__Interval_new5(const QMediaTimeRange__Interval* param1) {
	 return new QMediaTimeRange::Interval(*param1);
}

void QMediaTimeRange__Interval_CopyAssign(QMediaTimeRange__Interval* self, QMediaTimeRange__Interval* other) {
    *self = *other;
}

void QMediaTimeRange__Interval_MoveAssign(QMediaTimeRange__Interval* self, QMediaTimeRange__Interval* other) {
    *self = std::move(*other);
}

long long QMediaTimeRange__Interval_Start(const QMediaTimeRange__Interval* self) {
	return self->start();
}

long long QMediaTimeRange__Interval_End(const QMediaTimeRange__Interval* self) {
	return self->end();
}

bool QMediaTimeRange__Interval_Contains(const QMediaTimeRange__Interval* self, long long time) {
	return self->contains(time);
}

bool QMediaTimeRange__Interval_IsNormal(const QMediaTimeRange__Interval* self) {
	return self->isNormal();
}

QMediaTimeRange__Interval* QMediaTimeRange__Interval_Normalized(const QMediaTimeRange__Interval* self) {
	return new QMediaTimeRange::Interval(self->normalized());
}

QMediaTimeRange__Interval* QMediaTimeRange__Interval_Translated(const QMediaTimeRange__Interval* self, long long offset) {
	return new QMediaTimeRange::Interval(self->translated(offset));
}

void QMediaTimeRange__Interval_Delete(QMediaTimeRange__Interval* self) {
    delete self;
}

